proposals/0529-th-quotes-as-patterns.rst

```diff
@@ -0,0 +1,356 @@
+
+Costs and Drawbacks
+-------------------
+
+Action asymmetry
+~~~~~~~~~~~~~~~~
+
+The lack of symmetry where expression create actions but patterns only bind plain AST values is annoying.
+But the fixes for this might be too radical?
+
+Alternatives
+------------
+
+Pattern match on actions(!)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+I hypothesize that we could do better than the proposed actions vs no action asymmetry by meditating on the ways pattern matching relates to optics.
+For example::
+
+  [| a + $(x) |]
+
+could match any ``t Exp`` where ``t`` is an bind ``x :: t Expr`` with this desugaring::
+
+  f [| a + $(x) |] = ...
+
+  =>
+
+  f (Just x <- traverse __inner) = ...
+
+  __inner (AppE (VarE __a) x) | __a == 'a = Just x
+  __inner _                               = Nothing
+
+This behavior seems overwrought, as we are doing the as-proposed behavior *plus* an additional traversal.
+But this matches the fact that expression-position quotes are do what idiom brackets do (implicit ``Applicative``) in addition to base quoting.
+
+Support local variables
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Quotes that bind local variables do in fact have an interpretation as non-linear patterns::
+
+  f [| \x -> x |] = ...
+
+  =>
+
+  f (LamE [VarP __x0] (VarE __x1) | nameBase __x0 == "x" && __x0 == __x1 = ...
+
+This gets especially interesting with multiple scopes::
+
+  f [| (\x -> x, \x -> x) |] = ...
+
+  =>
+
+  f (TupE [ Just (LamE [VarP __x0] (VarE __x1))
+          , Just (LamE [VarP __x2] (VarE __x3))
+          ])
+    | nameBase __x0 == "x" && __x0 == __x1
+    | nameBase __x0 == "x" && __x2 == __x3
+    = ...
+
+Note how ``__x0`` is related to ``__x1`` and  ``__x2`` likewise to ``__x3``, but the former two are *not* related to the latter two.
+This respects the two independent scopes.
+
+This is perhaps convenient, but it rather baroque.
+It is also unclear whether the ``nameBase _ == "x"`` is useful in practice, or whether it is better to just "bake in" alpha equivalence and not care whether the local variable is an "x" or not provided the usage lines up with the binding.
+
+Finally, the non-linear patterns trick is not a true dual because it merely checks whether the variables "ended up" being the same *once the action is run*, rather than pattern matching on the action *itself*.
+De Bruijn indices encode actions in a way that makes equality of the easily decidable, for example, and thus would be a less hacky solution.
+
+---------
+
+Both alternatives are tempting, but I rather wait for more research on patterns, optics, and "dualizing" ``Applicative`` and effects like ``Quote`` in general, so we can better understand the theory of what's going on.
+If and when we understand the lay-of-the-land better, we can make some new extensions and deprecate the old ones accordingly.
+
+Unresolved Questions
+--------------------
+
+None at this time.
+
+Related Work
```

Comment:
I am frankly not sure I am expert enough either.

GHC is now quite beyond the state of the art in hygiene theory (probably Nominal Sets) or practice (probably Racket).

I am trying hard to straight-jacket the system so that the underlying lack of binding structure doesn't shine through, but it isn't entirely clear to me to what extend my efforts succeed.

`[A]X` Nominal set name bindings can I believe be related to `Quote`, in that  it should be possible to log all the "new names" and then establish an equivalence class the term with permutations of them, but:

- Once we run the action we loose that issue
- We will loose the *internal* binding structure because this is not compositional, with subterms having their own post-hoc reconstructable actions. Monad-ing everything together looses that information.

Reply:
I just stumbled upon https://www.mail-archive.com/racket-users@googlegroups.com/msg37427.html from @lexi-lambda, so I would say and she and @david-christiansen would both be good people with a foot in Haskell and Racket land to help expand this section.

Reply:
https://srfi.schemers.org/srfi-72/srfi-72.html is not quite formal, but interesting. I think @lexi-lambda in that thread and me reading now but found it appealing for much the same reasons.

Reply:
Correct me if I'm wrong, but since the proposal currently does not allow matching on binding constructs, we can't make hygiene better or worse. Judging from an hour worth reading, I also don't think we lock ourselves out with this proposal to more modern approaches such as `syntax-parse` because the stated semantics should be a subset of whatever advanced stuff we will happen to want in the future.